cora/signal/lss.py

```diff
@@ -1379,3 +1418,154 @@ def za_density_sph(
+    niter : int
+        Number of realizations to generate.
+    """
+
+    nside = config.Property(proptype=int, default=512)
+    frequencies = config.Property(proptype=lssutil.linspace, default=None)
+    full_pol = config.Property(proptype=bool, default=True)
+    variance = config.Property(proptype=float, default=None)
+    P_SN = config.Property(proptype=float, default=None)
+    use_freq_dependent_voxel_volume = config.Property(proptype=bool, default=False)
+    niter = config.Property(proptype=int, default=1)
+
+    def setup(self):
+        """Set up task."""
+        if ((self.variance is None) and (self.P_SN is None)) or (
+            (self.variance is not None) and (self.P_SN is not None)
+        ):
+            raise ValueError("Only one of variance or P_SN can be specified.")
+
+    def process(self) -> Map:
+        """Generate a flat-spectrum noise-like sky map with given power.
+
+        Returns
+        -------
+        out_map
+            Output Map container.
+        """
+
+        # Form frequency map in appropriate format to feed into Map container
+        freq = self.frequencies
+        nfreq = len(freq)
+        redshift = units.nu21 / freq - 1
+        freqmap = np.zeros(nfreq, dtype=[("centre", np.float64), ("width", np.float64)])
+        freqmap["centre"][:] = freq[:]
+        freqmap["width"][:] = np.abs(np.diff(freq[:])[0])
+
+        # Find index of central channel
+        ref_chan = int(nfreq / 2.0)
+
+        # Compute voxel volume
+        omega = healpy.nside2pixarea(self.nside)  # Pixel area in sr
+        if self.use_freq_dependent_voxel_volume:
+            dV = differential_comoving_volume(redshift)
+            dz = lssutil.calculate_width(redshift)
+        else:
+            dV = differential_comoving_volume(redshift[ref_chan])
+            dz = redshift[ref_chan + 1] - redshift[ref_chan]
+        voxvol = dV * dz * omega
+
+        # Make new map container and distribute in pixel, to make it easier to
+        # fill map with freq- and pol-dependent values
+        m = Map(freq=freqmap, polarisation=self.full_pol, nside=self.nside)
+        m.redistribute("pixel")
+
+        # Set standard deviation of map values
+        if self.variance is not None:
+            scale = self.variance**0.5
+        else:
+            scale = self.P_SN**0.5
+            if self.use_freq_dependent_voxel_volume:
+                scale /= voxvol[:, np.newaxis] ** 0.5
+            else:
+                scale /= voxvol**0.5
+
+        # Fill local section of map with random values
+        m.map[:].local_array[:, 0, :] = self.rng.normal(
+            scale=scale, size=(nfreq, m.map[:].local_shape[-1])
+        )
+
+        # Save voxel volume(s) as attribute
+        m.attrs["voxvol_ref"] = voxvol
+        m.attrs["central_redshift"] = redshift[ref_chan]
+
+        # Distribute in frequency, because other tasks often expect this
+        m.redistribute("freq")
+
+        if self._count == self.niter:
```

Comment:
This, along with `self.niter`, isn't necessarily needed. There's a built-in parameter called `limit_outputs` that's defined on the base task class in `caput` and tracks the number of times the task produced an output, so that can just be set to 1 in the pipeline config file and we don't have to track `self._count` here. However, the default would be different (here it's 1, wheres `limit_outputs` is set to produce unlimited outputs by default), so it's up to you to decide on which implementation to use.

Reply:
I'll note as well that I'm eventually planning to tidy up that internal task code

Reply:
Yeah, it could be useful as a simple way to generate multiple realizations. I've changed `niter` to `num_sims` to make that clearer.

Reply:
(Thanks for writing about `limits_outputs` though - I wasn't aware of that option before.)